google_compute_engine_oslogin/pam_module/pam_oslogin_login.cc

```diff
@@ -58,9 +58,9 @@ PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
   url << kMetadataServerUrl << "users?username=" << UrlEncode(str_user_name);
   string response;
   long http_code = 0;
-  if (!HttpGet(url.str(), &response, &http_code) || http_code == 500) {
-    // First check if there is a local file indicating that this is an oslogin
-    // user that has logged in before.
+  // Check for server failures. In this case, we can check for a local file to
+  // see if the user is an oslogin user.
+  if (!HttpGet(url.str(), &response, &http_code) || http_code >= 500) {
```

Comment:
Hmm, when we get a client error (anything between 201 to 499), it's probably a **programming bug**. But regardless, we still failed to get the user info from the server, and can't reliably tell if the user in question is an os-login user or not. In that case, shouldn't we fall back the to local user file as we do in the case of server errors?

In the current implementation, when we get a 3xx error for example, we'll skip the both if checks and go directly to `ParseJsonToEmail`, which will result in a crash: getting 3xx doesn't necessarily mean `response` will be empty, right?

So I think a more reliable logic should be:

* If we get a 200 with a non-empty response, great, it's an os-login user, go ahead and check its login permission
* If we get a 404, great still, it's *not* an os-login user, return PAM_SUCCESS
* In all other cases, we can't reliably tell, fall back to checking the local user's file

WDYT?